editoast/src/models/paced_train.rs

```diff
@@ -126,16 +129,75 @@ impl PacedTrain {
         }
     }
 
-    pub fn iter_occurrences(&self) -> impl Iterator<Item = TrainSchedule> {
-        let base_occurrence = self.clone().into_train_schedule();
-
-        (0..self.num_occurrences()).map(move |occurrence_idx| TrainSchedule {
-            start_time: base_occurrence.start_time + self.interval * occurrence_idx as i32,
-            ..base_occurrence.clone()
+    fn find_modified_exception(&self, occurrence_idx: usize) -> Option<&PacedTrainException> {
+        self.exceptions.iter().find(|exception| {
+            matches!(
+                exception.exception_type,
+                ExceptionType::Modified { occurrence_index }
+                    if occurrence_index == occurrence_idx as i32 && !exception.disabled
+            )
         })
     }
 
-    pub fn num_occurrences(&self) -> usize {
+    pub fn iter_occurrences(&self) -> impl Iterator<Item = (TrainId, TrainSchedule)> {
+        let created_exceptions_train_schedules = self
+            .exceptions
+            .iter()
+            .filter(|exception| matches!(exception.exception_type, ExceptionType::Created { .. }))
+            .map(|exception| {
+                (
+                    TrainId::PacedTrainCreatedException {
+                        paced_train_id: self.id,
+                        exception_key: exception.key.clone(),
+                    },
+                    self.apply_exception(exception),
+                )
+            });
+
+        let occurrences_train_schedules =
+            (0..self.num_base_occurrences()).map(move |occurrence_idx| {
+                let exception = self.find_modified_exception(occurrence_idx);
```

Comment:
This is not ideal sine we're iterating over the exception list for each number of occurrences.

Something we could do is create a table of `Option<&PacedTrainException>`.
Doing so make the `find_modified_exception` useless and then can be deleted.

```rust
let mut modified_exceptions = vec![None; self.num_base_occurrences()];
for exception in self.exceptions.iter() {
    if let ExceptionType::Modified { index, .. /* TODO */  } = exception {
        modified_exceptions[index] = Some(exception);
    }
}
```

Then you can access the exception simply accessing the `modified_exception[i]`;

Reply:
@flomonster I’ve made the modifications